packages/gatsby/cache-dir/ensure-page-component-in-bundle.js

```diff
@@ -0,0 +1,35 @@
+const didCallServer = new Set()
+
+const ensureComponentInBundle = chunkName =>
+  new Promise(resolve => {
+    if (!didCallServer.has(chunkName)) {
+      const req = new XMLHttpRequest()
+      req.open(`post`, `/___client-page-visited`, true)
+      req.setRequestHeader(`Content-Type`, `application/json;charset=UTF-8`)
+      req.send(JSON.stringify({ chunkName }))
+
+      didCallServer.add(chunkName)
+    }
+
+    // Tell the server the user wants to visit this page
+    // to trigger it compiling the page component's code.
+    const checkForBundle = () => {
+      // Check if the bundle is included and return.
+      if (process.env.NODE_ENV !== `test`) {
+        delete require.cache[
+          require.resolve(`$virtual/lazy-client-sync-requires`)
+        ]
+      }
+
+      const lazyRequires = require(`$virtual/lazy-client-sync-requires`)
+      if (lazyRequires.lazyComponents[chunkName]) {
+        resolve(lazyRequires.lazyComponents[chunkName])
+      } else {
+        setTimeout(checkForBundle, 100)
+      }
+    }
+
+    checkForBundle()
```

Comment:
do we need to consider that requested `chunkname` actually would never show up in the bundle and this would never resolve or reject?

because of a lot of async execution here (including networking) there still might be the cases where template might no longer be there (was removed since initial request)

We did hit problem like this Vlad when working on query on demand in this test scenario: https://github.com/gatsbyjs/gatsby/blob/6858f22fc9a64f71756d5cd5122fab99d12749cf/e2e-tests/development-runtime/cypress/integration/navigation/redirect.js#L81-L112

which removes the page and immediately tries to visit it so page is still in store ( we didn't receive event from chokidar yet to remove the page), then pages get removed while we do wait for `page-data` for it - I think this is quite similar case?

Reply:
I think it's not blocking merge because this is experimental - but something to keep in mind?

Reply:
yeah... actually timing out makes a lot of sense — we can check for say 5 seconds and then quit. Otherwise the runtime will just keep looping forever trying to load it which is wasteful at best and could slow things down.